TGL machine for max98373_RT5682

sathya-nujella opened the pull request:
TGL Chrome DUT has the below codecs:
	Max98373 spk amp
	ALC5682 headset codec
	PDM DMIC

To support this,
	(a) Instead of adding a new mach file, re-using sof-rt5682 and added needed configuration
	(b) Adding a separate file for Max98373 for future easy re-use

plbossart commented on sound/soc/intel/boards/sof_maxim_common.h:
```diff
@@ -0,0 +1,88 @@
+	int j;
+
+	for (j = 0; j < runtime->num_codecs; j++) {
+		struct snd_soc_dai *codec_dai = runtime->codec_dais[j];
```
I think there's a for_each_ macro for this?

sathya-nujella replied:
Updated to use for_each_rtd_codec_dai.  Thank you!

sathya-nujella replied:
I see that for_each_rtd_codec_dais is new one. So I corrected and updated in newer patchset.

plbossart commented on sound/soc/intel/boards/sof_maxim_common.h:
```diff
@@ -0,0 +1,88 @@
+		}
+		if (!strcmp(codec_dai->component->name, MAX98373_DEV1_NAME)) {
+			/* DEV1 tdm slot configuration */
+			snd_soc_dai_set_tdm_slot(codec_dai, 0xC0, 3, 8, 16);
```
why do you need 8 slots, looks like you could use only 4?

sathya-nujella replied:
We wan to keep 8-TDM slots for the support if 4-amp's connected. So, same can be re-used.
Also, one advantage from Codec side is:
With 8-slots, amp had been verified well in KBL products.. so want to use same working configuration (which seems to be good for TGL too).

plbossart replied:
ok, but then look at sof_da7219_max98373, the code uses 4 slots.

sathya-nujella replied:
I cross checked, in that program there is no plan to go for 4-amp's.. so 4-slots is good for them. With these codec combinations in current projects, there is future plan for 4-amp's: so: I am keeping as 8-slots.

plbossart commented on sound/soc/intel/common/soc-acpi-intel-tgl-match.c:
```diff
@@ -96,6 +101,14 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_machines[] = {
+		.machine_quirk = snd_soc_acpi_codec_list,
+		.quirk_data = &tgl_hp_5682_codec,
+		.sof_fw_filename = "sof-tgl.ri",
+		.sof_tplg_filename = "sof-tgl-max98373-rt5682.tplg",
```
somehow gitHub lost my comment.

I think we should flip the logic and use rt5682 as a primary key and the amp as a quirk

sathya-nujella replied:
updated now.

ranj063 commented on sound/soc/intel/boards/sof_maxim_common.h:
```diff
@@ -0,0 +1,88 @@
+};
+
+static struct snd_soc_dai_link_component max98373_components[] = {
+	{  /* For Light */
```
Light?

sathya-nujella replied:
sorry, that is typo.
Corrected now.

sathya-nujella commented:
Hi,
For the new file addition, I am seeing this checkpatch warning.
         >WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
          >new file mode 100644
Can you please suggest me how I can avoid this warning?

lyakh commented on sound/soc/intel/boards/sof_maxim_common.h:
```diff
@@ -0,0 +1,87 @@
+	/* speaker */
+	{ "Left Spk", NULL, "Left BE_OUT" },
+	{ "Right Spk", NULL, "Right BE_OUT" },
+};
```
This doesn't feel quite right... Defining static data in a header means every other file `.c` or `.h` that includes this header will have identical instances of these objects... I understand that might indeed be what you want, because the goal of this header is to only be included by files, that need those instances, but this is somewhat unsafe (it's easy to overlook them and include the header in another file just to get struct defines or similar. I think a better idea would be to define macros here for these objects and to then instantiate them in each `.c` file separately

sathya-nujella replied:
> This doesn't feel quite right... Defining static data in a header means every other file `.c` or `.h` that includes this header will have identical instances of these objects... I understand that might indeed be what you want, because the goal of this header is to only be included by files, that need those instances, but this is somewhat unsafe (it's easy to overlook them and include the header in another file just to get struct defines or similar. I think a better idea would be to define macros here for these objects and to then instantiate them in each `.c` file separately

ok, I can move the function definitions to sof_maxim_common.c and will include sof_maxim_common.o in Makefile for the needed one. In the header file, let me remove static.
Does this method ok?

kv2019i replied:
That's a valid point @lyakh We don't have any good examples in our boards file now. 

In skl_hda_dsp_common.[hc] I used following:

header:
extern struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS];

c-file for common part:
struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS] = { // actual definition }

... but this a bit annoying if you don't need the implementation file for anything else.

Maybe macros would be cleaner still? Any good examples in kernel tree @lyakh ?

plbossart replied:
yes we can use modules, but we'd want to use namespaces so that the maxim_common symbols are only used from specific machine drivers.

lyakh replied:
@kv2019i examples depend of how exactly those objects should be used. From this implementation it looks like every user would have exactly the same object definitions - same names, same contents. But some of them aren't `const`. Does that mean that they really can be modified at run-time? I just checked, it wouldn't be possible to make at least some of them `const`. Still, if they aren't really going to change, they can be placed in a common .c. If they can indeed change and every user should get a private copy of them, then we could use macros similar to those in soc-dapm.h.

sathya-nujella replied:
>but this is somewhat unsafe (it's easy to overlook them and include the header in another file just to get struct defines or similar.
Hi @lyakh ,
I have now updated in new patchset ..only keeping needed in header file and moved others to sof_maxim_common.c

kv2019i commented on sound/soc/intel/common/soc-acpi-intel-tgl-match.c:
```diff
@@ -79,6 +79,11 @@ static const struct snd_soc_acpi_link_adr tgl_rvp[] = {
 	{}
```
For the commit message.  The indentation in commit message seems a bit of, maybe just use: 
```
ASoC: Intel: common: add mach entry for tgl_max98373_rt5682

Update tgl mach table with: Maxim98373 Amp and ALC5682 hp codec. Both 
of the codecs are on I2S bus.
```

sathya-nujella replied:
Thank you @kv2019i for the suggestion. I have updated now.

sathya-nujella commented:
> Hi,
> For the new file addition, I am seeing this checkpatch warning.
> >WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
> >new file mode 100644
> Can you please suggest me how I can avoid this warning?

Ranjani told this does not need to be fixed, I got an answer. Thanks!

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,77 @@
+#include <sound/soc.h>
+#include "sof_maxim_common.h"
+
+const struct snd_soc_dapm_route max_98373_dapm_routes[] = {
```
static

sathya-nujella replied:
Added static. Updated patches now.

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,77 @@
+	{ "Right Spk", NULL, "Right BE_OUT" },
+};
+
+struct snd_soc_codec_conf max_98373_codec_conf[] = {
```
static

sathya-nujella replied:
Added static. Updated patches now.

plbossart commented on sound/soc/intel/boards/sof_maxim_common.h:
```diff
@@ -0,0 +1,26 @@
+#define MAX_98373_DEV1_NAME	"i2c-MX98373:01"
+
+extern const struct snd_soc_dapm_route max_98373_dapm_routes[];
+extern struct snd_soc_codec_conf max_98373_codec_conf[];
```
these two can be hidden as static, and since you use ARRAY_SIZE they should definitively be static

sathya-nujella replied:
Thank you. Removed them and added static for those in c file. Updated patches now.

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -745,6 +753,11 @@ static int sof_audio_probe(struct platform_device *pdev)
 
+	if (sof_rt5682_quirk & SOF_MAX98373_SPEAKER_AMP_PRESENT) {
+		sof_max98373_codec_conf(&sof_audio_card_rt5682);
+		sof_audio_card_rt5682.name = "tgl_max73rt82";
```
why the tgl prefix?
and why change the card name anyways?

sathya-nujella replied:
>why the tgl prefix?
I will remove tgl prefix, add sof_. I will update patches.

> why change the card name anyways?
We are using one OS to support multiple Audio daughter boards (i.e. multiple audio configurations like: TGL with max_98373_rt5682, with max_98357A_rt5682 etc tec). CRAS config folder name is same as sound card name. CRAS config folder will have tuning data and ucm files. They should be unique to have one OS support. So, want to change card name to support relevant audio daughter card/configuration.

plbossart replied:
@sathya-nujella the direction we had from Jaroslav was to use strings that can be retrieved by userspace with 'control components', e.g. spk:max98373 hs:rt5682 so that the card name remains the same instead of being an encoding of the platform configuration

see e.g. commits
0d5c8187562848b619a35f2ffc5e18ce703e9f3d b5706f8ec29fb461571d25e3e813ede121fe31cd 
8cd9956f61c65022209ce6d1e55ed12aea12357d

I am planning on using this for SoundWire, maybe we should do the same here.

@cujomalainey @dgreid FYI on UCM impact

sathya-nujella replied:
We will work with Google to get components_string to be used in CRAS side.. may be I will raise a public chromium and work with Dylan/Curtis.
In new patchset, I updated card name as needed now.
As soon as user space support comes, I can add PR to add components_string and remove this overriding. Please suggest if this is ok to continue until full working frame work is available.

plbossart replied:
@sathya-nujella I suggest you put this name change in a different patch, and we keep the name change in the SOF tree until the rest is complete, while we upstream the rest of your changes. That way we only upstream stuff that works with the component part. Would this work for you?

sathya-nujella replied:
sure @plbossart . Thank you for suggesting alternative ideas.
I have updated patchset now and move to a separate patch only this overriding the card name.

plbossart replied:
@sathya-nujella I just started doing this for the SoundWire card, and that would give something like:

````
root@plb:~# more /sys/class/sound/card0/components 
HDA:8086280b,80860101,00100000 cfg-spk:2 hs:rt711 spk:rt1308 mic:rt715
````
userspace can query/parse that string w/ alsa-lib and know what components are handled by the card. It's a lot more flexible/scalable than using a card name :-)

sathya-nujella replied:
> @sathya-nujella I just started doing this for the SoundWire card, and that would give something like:
> 
> ```
> root@plb:~# more /sys/class/sound/card0/components 
> HDA:8086280b,80860101,00100000 cfg-spk:2 hs:rt711 spk:rt1308 mic:rt715
> ```
> 
> userspace can query/parse that string w/ alsa-lib and know what components are handled by the card. It's a lot more flexible/scalable than using a card name :-)

Sure Pierre. Understood..Thank you!

cujomalainey replied:
Ack, so this means that all cards using the same codecs will bear the same name? Will we still be able to use the card name or will have to change to the spk: hs: notation for the ucm?

sathya-nujella replied:
sound card name will be same "sof-rt5682" for all machines which ever uses this machine driver "sof-rt5682.c". For CRAS to have unique ucm/cras-config files, machine drivers should set"card.components" filed. component field with "cfg-spk:<> spk:<> cfg-hp:<hp> cfg-mic/cfg-dmics:". Based on this CRAS has to re-construct a string to load ucm files. If this component field is not set, then CRAS will have to load regular card name based ucm files.

I summarized based on my understanding.. @plbossart @cujomalainey : please see if this is line with expectation.

cujomalainey replied:
@sathya-nujella thanks for the explanation. With unibuild this should not be an issue since we can install per unit ucms by redirecting cras with suffixes to the ucms.

sathya-nujella replied:
Thank you @cujomalainey . Based on this, I will drop patch:3 and we do not need to keep snd card name overriding patch.

sathya-nujella commented:
> LGTM. If you can add a comment for maintainers that the 3rd patch should not go upstream it'd be even nicer. 
Done. Added [NOT FOR UPSTREAM] tag for 3rd patch.

lyakh commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,77 @@
+#include <sound/soc.h>
+#include "sof_maxim_common.h"
+
+static const struct snd_soc_dapm_route max_98373_dapm_routes[] = {
```
strictly speaking for this you have to `#include <sound/soc-dapm.h>`

sathya-nujella replied:
Done. I have updated patchset.
@lyakh : please review.

lyakh replied:
no, sorry, that isn't what I meant. The standard approach in the kernel is to include all headers where they are needed.

sathya-nujella replied:
I have included all snd header files need. Please review now. Thanks!

lyakh commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,77 @@
+};
+
+int max98373_hw_params(struct snd_pcm_substream *substream,
+		       struct snd_pcm_hw_params *params)
```
also you'd need to `#include <uapi/sound/asound.h>` or at least `#include <sound/asound.h>` but most seem to just `#include <sound/pcm.h>` and rely on it including sound/asound.h...

sathya-nujella replied:
Done. I have updated patchset.

lyakh commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,77 @@
+		       struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai;
```
`#include <sound/soc-dai.h>`

sathya-nujella replied:
Done. I have updated patchset.

lyakh commented on sound/soc/intel/boards/sof_maxim_common.h:
```diff
@@ -0,0 +1,24 @@
+int max98373_spk_codec_init(struct snd_soc_pcm_runtime *rtd);
+void sof_max98373_codec_conf(struct snd_soc_card *card);
+int max98373_hw_params(struct snd_pcm_substream *substream,
+		       struct snd_pcm_hw_params *params);
```
need to forward-declare all the structs here

sathya-nujella replied:
Done. I have updated patchset.

ranj063 commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,80 @@
+#include <sound/soc.h>
+#include <sound/soc-dai.h>
+#include <sound/soc-dapm.h>
+#include <uapi/sound/asound.h>
```
@sathya-nujella you can remove the ones that are already included in sof_maxim_common.h

sathya-nujella replied:
Hi @ranj063 
If I partially move the header files.. I see build errors.
So, I moved all sound include files to sof_maxim_common.h and its c file only includes it.

lyakh replied:
I wouldn't do that. Every .c and .h file has to include its required headers itself without relying on recursive inclusion.

sathya-nujella replied:
I went through all lines and identified needed snd headers and included them now. Please review now. Thanks!

sathya-nujella replied:
Hi @lyakh ,
For my learning, I want to understand this:
1) >Every .c and .h file has to include its required headers itself without relying on recursive inclusion.
Let's say in my case, if I put all snd headers in sof_maxim_common.h and sof_maxim_common.c only includes /sof_maxim_common.h.
Build wise and functionality wise it works. For my learning, asking... what kind of issue comes because of that maintainers ask to include snd headers differently like every *.c and *.h needs to include them.. 

2) regarding forwarding defining structures:
Again for my understanding.. let's say: In sof_maxim_common.h, I include related snd headers.. then we don't need to do forward define them, right?
Forward define is only needed if those header files are not included..

Please share inputs on this.

lyakh replied:
@sathya-nujella answering to above points:
1. Not everything that works is also a good way to do it. My first pascal program - I had to print it out on paper back then - was basically filling each line from position 0 to position 79... (and mind you, it was a long and rather complex program). We don't do that today. Including all required headers in the file is basically documentation and ease of reading. You document where various parts come from and the reader has to only do one jump to respective declarations instead of following potentially multi-level nested headers. Imagine, if you include 3 headers in your .c. Each of those headers collects other headers - say, sound, pm, and others. But those headers also rely on recursive inclusion, so, to find a definition of `printk()` you first have to find a complete path through all the recursions. With this requirement it's always just one jump.
2. Sure, if you include a header with a complete declaration of a struct, you don't need to forward-declare it.

sathya-nujella replied:
Thanks a lot @lyakh for explaining in detail. Understood now.

lyakh commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,75 @@
+	int j;
+
+	for_each_rtd_codec_dais(rtd, j, codec_dai) {
+		if (!strcmp(codec_dai->component->name, MAX_98373_DEV0_NAME)) {
```
`#include <linux/string.h>` for that

sathya-nujella replied:
Done. updated now.

lyakh commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,75 @@
+	},
+};
+
+int max98373_hw_params(struct snd_pcm_substream *substream,
```
can this function be `static`?

sathya-nujella replied:
Done. Updated now.

lyakh commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -0,0 +1,80 @@
+	ret = snd_soc_dapm_add_routes(&card->dapm, max_98373_dapm_routes,
+				      ARRAY_SIZE(max_98373_dapm_routes));
+	if (ret)
+		dev_err(rtd->dev, "Speaker map addition failed: %d\n", ret);
```
I'll mark it as optional but strictly speaking you also need `linux/device.h` for `dev_err()` (or `linux/dev_printk.h` if you prefer...) and `linux/kernel.h` for `ARRAY_SIZE()`.

sathya-nujella replied:
Sure @lyakh , wll keep this info and take care next time. Thanks!

kv2019i commented:
SOFCI TEST
